Use IOSurface for texture sharing on macOS

zakorgy opened the pull request:
<!-- Please describe your changes on the following line: -->

Use a non-shared GL context for the WebGL threads, with triple buffered IOSurfaces bound to the Framebuffer. Using a simple IOSurface resulted in flickering since both the WR and the WebGL thread uses it, using three gave me the proper result.
When we receive a `Lock` message, we send the id of a finished IOSurface to the WR thread where we bind it to a texture, which WR can use.
The synchronization of the IOSurfaces is a little hacky, since there are no clue in GL that we started rendering a new frame and we can swap the surfaces: if the document process a `FrameRequestCallback` callback (which comes from `requestAnimationFrame`), we send a message to the WebGL thread to notify about the start of the new frame. This is where we swap the underlying IOsurfaces of the WebGL contexts.
I have checked the performance on several WebGL samples and the FPS numbers where promising.

---
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `___` with appropriate data: -->
- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] These changes fix #11138 (GitHub issue number if applicable)

<!-- Also, please make sure that "Allow edits from maintainers" checkbox is checked, so that we can help you if you get stuck somewhere along the way.-->

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/23509)
<!-- Reviewable:end -->

Commit:
Use texture sharing with IOSurfaces on macOS

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/document.rs, components/script/dom/window.rs
 * @kichjang: components/script/dom/document.rs, components/script/dom/window.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify script code, but no tests are modified. Please consider adding a test!

zakorgy commented:
cc @jdm

asajeffrey commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -70,18 +72,29 @@ impl WebGLThreads {
-        WebGLSender<(u32, Size2D<i32>, usize)>,
-        WebGLReceiver<(u32, Size2D<i32>, usize)>,
+        WebGLSender<(u32, Size2D<i32>, Option<u32>, usize)>,
+        WebGLReceiver<(u32, Size2D<i32>, Option<u32>, usize)>,
```
Ditto: any chance we can name this type and name its fields?

asajeffrey commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -70,18 +72,29 @@ impl WebGLThreads {
     webrender_gl: Rc<dyn gl::Gl>,
     webgl_channel: WebGLSender<WebGLMsg>,
+    // Mapping between an IOSurface and the texture it is bound on the WR thread
+    textures: FnvHashMap<u32, gl::GLuint>,
```
Nit: any chance we can name these types? I'm getting lost in all the u32s :)

asajeffrey commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -94,13 +107,34 @@ impl WebGLExternalImageApi for WebGLExternalImages {
-        (image_id, size)
+        let (image_id, size, io_surface_id, gl_sync) = self.lock_channel.1.recv().unwrap();
+
+        let texture_id = match io_surface_id {
```
Nit: a comment about what's going on here would be useful.

asajeffrey commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -94,13 +107,34 @@ impl WebGLExternalImageApi for WebGLExternalImages {
+        let texture_id = match io_surface_id {
+            Some(io_surface_id) => {
+                let texture_id = match self.textures.entry(io_surface_id) {
+                    Entry::Occupied(o) => *o.get(),
```
Nit: can we use `Entry::or_insert_with()` here?

zakorgy replied:
As I remember we can't because of the borrow of `self` [here](https://github.com/servo/servo/pull/23509/files/055fafb452aa053d520215949100d3a62a436aaa#diff-250070c6c5a38c7f9fa0f5b3c101f68bR118), but I will check it again.

asajeffrey commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -94,13 +107,34 @@ impl WebGLExternalImageApi for WebGLExternalImages {
+                        *v.insert(texture_id)
+                    },
+                };
+                texture_id
```
There's no locking or sync'ing needed here? How does this avoid race conditions between the WebGL thread and the WebRender thread?

zakorgy replied:
So far the triple buffering result worked for me, because the time interval between `Lock` and `Unlock` messages was lesser than switching texture targets on the WebGL thread two times. But I can agree adding it would be make the code more safer, I will look into that.

asajeffrey commented on components/canvas/webgl_thread.rs:
```diff
@@ -189,13 +192,32 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
 
+    /// Swap the underlying IOsurfaces upon requestAnimationFrame is called
+    fn swap_draw_buffers(&mut self, sender: WebGLSender<()>) {
+        for (context_id, ref mut data) in self.contexts.iter_mut() {
```
It would be nice if we could avoid using two different mechanisms (triple-buffering on MacOS and GL sync'ing on other OSs). Is there not a way to use the existing locking API with IOSurfaces?

asajeffrey commented on components/canvas/webgl_thread.rs:
```diff
@@ -277,18 +312,33 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
             .new_shared_context(version, size, attributes)
             .map(|r| (r, WebGLContextShareMode::SharedTexture))
+            .or_else(|err| {
+                #[cfg(target_os = "macos")]
```
Should we introduce a new abstraction so we're not special-casing `macos`? Something where the GL context isn't sharing, but the texture is? Alternatively, should we move this logic into `GLFactory`?

asajeffrey reviewed (commented):
Yay! It would be very very nice to have this. I'm happily surprised we can do this without any changes to webrender! My main question is about whether we can do this but keep the existing sync'ing mechanism, so we don't end up with very different stories of how to avoid race conditions on different architectures.

zakorgy commented:
@asajeffrey sorry for the late answer and thank you for the review. For answering your main question, I tried to use the current syncing mechanism, but unfortunately this was the only satisfying solution I could come up.

asajeffrey commented:
Oh that's annoying. We should be using some sync mechanism rather than relying on timing. Perhaps we can use a variant on triple buffering, where we keep 3 buffers: the one that webrender is reading from, the one the webgl is currently writing to, and the one that webgl has finished writing to? Lock would swap 1 and 3, and at the end of an rAF we'd flush and swap 2 and 3?

zakorgy commented:
@asajeffrey your idea works well, I will address the review changes and update the PR. After that I will check if we could abstract away the cfg-s in `GLFactory` as you mentioned in https://github.com/servo/servo/pull/23509#discussion_r290391599

jdm commented:
Oh, good find on using the correct clear color!

asajeffrey commented on Cargo.toml:
```diff
@@ -28,3 +28,4 @@ opt-level = 3
 # Those are here to dedupe winapi since mio is still using winapi 0.2.
 mio = { git = "https://github.com/servo/mio.git", branch = "servo" }
 iovec = { git = "https://github.com/servo/iovec.git", branch = "servo" }
+offscreen_gl_context = { git = "https://github.com/zakorgy/rust-offscreen-rendering-context", branch = "io-surface" }
```
We should land this patch to `offscreen_gl_context` before landing this PR.

jdm commented:
@bors-servo try=wpt-mac

bors-servo commented:
:hourglass: Trying commit 5a752174bc050efd2b9a688283529e3cf466b654 with merge d05a4142cc34536ca1cbdd6bbc1fec66edc37b7c...

jdm commented:
Also the linux and windows builds are failing:
```
error[E0599]: no method named `get_complete_io_surface_id` found for type `&offscreen_gl_context::draw_buffer::DrawBuffer` in the current scope
   --> components\canvas\gl_context.rs:182:37
    |
182 |                         draw_buffer.get_complete_io_surface_id(),
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
error[E0599]: no method named `get_complete_io_surface_id` found for type `&offscreen_gl_context::draw_buffer::DrawBuffer` in the current scope
   --> components\canvas\gl_context.rs:196:37
    |
196 |                         draw_buffer.get_complete_io_surface_id(),
    |                                     ^^^^^^^^^^^^^^^^^^^^^^^^^^
error[E0599]: no method named `get_active_io_surface_id` found for type `&offscreen_gl_context::draw_buffer::DrawBuffer` in the current scope
   --> components\canvas\gl_context.rs:210:51
    |
210 |                 ctx.borrow_draw_buffer().unwrap().get_active_io_surface_id()
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^
error[E0599]: no method named `get_active_io_surface_id` found for type `&offscreen_gl_context::draw_buffer::DrawBuffer` in the current scope
   --> components\canvas\gl_context.rs:213:51
    |
213 |                 ctx.borrow_draw_buffer().unwrap().get_active_io_surface_id()
    |                                                   ^^^^^^^^^^^^^^^^^^^^^^^^
error[E0599]: no method named `swap_draw_buffer` found for type `&mut offscreen_gl_context::gl_context::GLContext<offscreen_gl_context::platform::with_wgl::native_gl_context::NativeGLContext>` in the current scope
   --> components\canvas\gl_context.rs:222:58
    |
222 |             GLContextWrapper::Native(ref mut ctx) => ctx.swap_draw_buffer(clear_color),
    |                                                          ^^^^^^^^^^^^^^^^
error[E0599]: no method named `swap_draw_buffer` found for type `&mut offscreen_gl_context::gl_context::GLContext<offscreen_gl_context::platform::with_osmesa::OSMesaContext>` in the current scope
   --> components\canvas\gl_context.rs:223:58
    |
223 |             GLContextWrapper::OSMesa(ref mut ctx) => ctx.swap_draw_buffer(clear_color),
    |                                                          ^^^^^^^^^^^^^^^^
error[E0599]: no method named `handle_lock` found for type `&mut offscreen_gl_context::gl_context::GLContext<offscreen_gl_context::platform::with_wgl::native_gl_context::NativeGLContext>` in the current scope
   --> components\canvas\gl_context.rs:229:58
    |
229 |             GLContextWrapper::Native(ref mut ctx) => ctx.handle_lock(),
    |                                                          ^^^^^^^^^^^
error[E0599]: no method named `handle_lock` found for type `&mut offscreen_gl_context::gl_context::GLContext<offscreen_gl_context::platform::with_osmesa::OSMesaContext>` in the current scope
   --> components\canvas\gl_context.rs:230:58
    |
230 |             GLContextWrapper::OSMesa(ref mut ctx) => ctx.handle_lock(),
    |                                                          ^^^^^^^^^^^
error: aborting due to 8 previous errors
```

jdm commented:
Lots of test failures - here are the unique ones:
```
▶ CRASH [expected OK] /_webgl/conformance/extensions/ext-shader-texture-lod.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ called `Option::unwrap()` on a `None` value (thread WebGLThread, at src/libcore/option.rs:347)
  │ stack backtrace:
  │    0:        0x112c9430e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x112c933ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x10ff70669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x112e40410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x112e3fe5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x112e3fd48 - _rust_begin_unwind
  │    6:        0x112e5cfe1 - __ZN4core9panicking9panic_fmt17h08425221a00ef4a4E
  │    7:        0x112e5cf26 - __ZN4core9panicking5panic17h964b30896190e4ffE
  │    8:        0x112414ae5 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    9:        0x1123de816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │   10:        0x112461561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   11:        0x112e48c4e - ___rust_maybe_catch_panic
  │   12:        0x112461771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   13:        0x112e2e7fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   14:        0x112e4812d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   15:     0x7fff76978660 - __pthread_body
  │   16:     0x7fff7697850c - __pthread_start
  │ [2019-06-11T16:49:29Z ERROR servo] called `Option::unwrap()` on a `None` value

▶ TIMEOUT [expected OK] /_webgl/conformance/buffers/buffer-data-dynamic-delay.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T16:59:41Z ERROR canvas::webgl_thread] Last GL operation failed: DrawArrays { mode: 5, first: 0, count: 2 }
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x1088c430e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x1088c33ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x105ba0669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x108a70410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x108a6fe5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x108a6fdb0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x10804c711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x108044bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x10800e816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x108091561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x108a78c4e - ___rust_maybe_catch_panic
  │   11:        0x108091771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x108a5e7fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x108a7812d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff76978660 - __pthread_body
  │   15:     0x7fff7697850c - __pthread_start

  ▶ CRASH [expected OK] /_webgl/conformance2/rendering/texture-switch-performance.html
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T17:02:09Z ERROR canvas::webgl_thread] Last GL operation failed: Uniform1i(0, 0)
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x10dc7730e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x10dc763ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x10af53669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x10de23410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x10de22e5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x10de22db0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x10d3ff711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x10d3f7bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x10d3c1816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x10d444561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x10de2bc4e - ___rust_maybe_catch_panic
  │   11:        0x10d444771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x10de117fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x10de2b12d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff584bd660 - __pthread_body
  │   15:     0x7fff584bd50c - __pthread_start

    ▶ CRASH [expected OK] /_webgl/conformance2/rendering/canvas-resizing-with-pbo-bound.html
  │ 
  │ _RegisterApplication(), FAILED TO establish the default connection to the WindowServer, _CGSDefaultConnection() is NULL.
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T16:56:42Z ERROR script::dom::webglrenderingcontext] Couldn't create WebGLRenderingContext: preserveDrawingBuffer is not supported yet
  │ DESCRIPTION: Verifies that resizing the canvas (recreating the backing framebuffer) works correctly while a PBO is bound.
  │ 
  │ Regression test for Chromium <a href='https://bugs.chromium.org/p/chromium/issues/detail?id=644572'>Issue 644572</a>
  │ 
  │ requestAnimationFrame
  │ Testing preserveDrawingBuffer = true
  │ FAIL Unable to fetch WebGL rendering context for Canvas
  │ FAIL context does not exist
  │ Testing preserveDrawingBuffer = false
  │ PASS context exists
  │ [2019-06-11T16:56:42Z ERROR canvas::webgl_thread] Last GL operation failed: ClearColor(0.0, 1.0, 0.0, 1.0)
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x10780030e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x1077ff3ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x104adc669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x1079ac410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x1079abe5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x1079abdb0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x106f88711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x106f80bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x106f4a816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x106fcd561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x1079b4c4e - ___rust_maybe_catch_panic
  │   11:        0x106fcd771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x10799a7fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x1079b412d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff584bd660 - __pthread_body
  │   15:     0x7fff584bd50c - __pthread_start

  ▶ CRASH [expected OK] /_webgl/conformance/context/context-no-alpha-fbo-with-alpha.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T17:08:49Z ERROR canvas::webgl_thread] Last GL operation failed: CreateFramebuffer(WebGLSender(..))
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x10af2730e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x10af263ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x108203669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x10b0d3410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x10b0d2e5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x10b0d2db0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x10a6af711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x10a6a7bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x10a671816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x10a6f4561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x10b0dbc4e - ___rust_maybe_catch_panic
  │   11:        0x10a6f4771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x10b0c17fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x10b0db12d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff58b01660 - __pthread_body
  │   15:     0x7fff58b0150c - __pthread_start

  ▶ CRASH [expected OK] /_webgl/conformance/context/context-hidden-alpha.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T17:08:48Z ERROR canvas::webgl_thread] Last GL operation failed: BlendFunc(0, 772)
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x10890230e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x1089013ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x105bde669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x108aae410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x108aade5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x108aaddb0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x10808a711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x108082bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x10804c816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x1080cf561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x108ab6c4e - ___rust_maybe_catch_panic
  │   11:        0x1080cf771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x108a9c7fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x108ab612d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff58b01660 - __pthread_body
  │   15:     0x7fff58b0150c - __pthread_start

  ▶ TIMEOUT [expected OK] /_webgl/conformance/rendering/many-draw-calls.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T16:49:17Z ERROR canvas::webgl_thread] Last GL operation failed: Clear(16384)
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x111ad930e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x111ad83ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x10edb5669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x111c85410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x111c84e5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x111c84db0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x111261711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x111259bb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x111223816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x1112a6561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x111c8dc4e - ___rust_maybe_catch_panic
  │   11:        0x1112a6771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x111c737fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x111c8d12d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff58df6660 - __pthread_body
  │   15:     0x7fff58df650c - __pthread_start

  ▶ CRASH [expected OK] /_webgl/conformance/rendering/scissor-rect-repeated-rendering.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ [2019-06-11T16:54:36Z ERROR canvas::webgl_thread] Last GL operation failed: SetViewport(64, 64, 64, 64)
  │ assertion failed: `(left == right)`
  │   left: `1282`,
  │  right: `0`: Unexpected WebGL error: 0x502 (1282) (thread WebGLThread, at components/canvas/webgl_thread.rs:1555)
  │ stack backtrace:
  │    0:        0x110f1b30e - __ZN9backtrace9backtrace5trace17h2aa3bfd5ea4aba44E
  │    1:        0x110f1a3ac - __ZN72_$LT$backtrace..capture..Backtrace$u20$as$u20$core..default..Default$GT$7default17hc9b000eeaacd89abE
  │    2:        0x10e1f7669 - __ZN5servo4main28_$u7b$$u7b$closure$u7d$$u7d$17h9d9b5b0a24be810cE
  │    3:        0x1110c7410 - __ZN3std9panicking20rust_panic_with_hook17h4d29a1b082e04079E
  │    4:        0x1110c6e5c - __ZN3std9panicking18continue_panic_fmt17h4a5eb546816972c4E
  │    5:        0x1110c6db0 - __ZN3std9panicking15begin_panic_fmt17h1bef9ef925c1eb3eE
  │    6:        0x1106a3711 - __ZN6canvas12webgl_thread9WebGLImpl5apply17h0d8005f53d521abcE
  │    7:        0x11069bbb7 - __ZN6canvas12webgl_thread21WebGLThread$LT$VR$GT$20handle_webgl_command17h53f4cd3731b70ca0E
  │    8:        0x110665816 - __ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h76a2e1f313bceaf9E.llvm.10547903177341280181
  │    9:        0x1106e8561 - __ZN3std9panicking3try7do_call17h150ae451a94dc9a7E.llvm.5012549804399729135
  │   10:        0x1110cfc4e - ___rust_maybe_catch_panic
  │   11:        0x1106e8771 - __ZN4core3ops8function6FnOnce40call_once$u7b$$u7b$vtable.shim$u7d$$u7d$17h6a94b1439869bf36E
  │   12:        0x1110b57fd - __ZN83_$LT$alloc..boxed..Box$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$A$GT$$GT$9call_once17h3ec79158893c158bE
  │   13:        0x1110cf12d - __ZN3std3sys4unix6thread6Thread3new12thread_start17h42de43c740cec52bE
  │   14:     0x7fff58df6660 - __pthread_body
  │   15:     0x7fff58df650c - __pthread_start
```

jdm commented:
One difference that occurs when running tests is that we run them under OSMesa.

bors-servo commented:
:broken_heart: Test failed - [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Khh4vbjVScOLM-83VDyCmA)

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23468) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented:
@bors-servo try

bors-servo commented:
:hourglass: Trying commit b92d40740b0505a06d2d5047c6c9d680aaf8e4ff with merge 97bc8203a645b5db9068fbccb66f81bd67880357...

bors-servo commented:
:broken_heart: Test failed - [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/12617)

jdm commented:
```
  ▶ CRASH [expected OK] /_webgl/conformance/extensions/ext-shader-texture-lod.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ 
  │ (servo:32203): GStreamer-WARNING **: External plugin loader failed. This most likely means that the plugin loader helper binary was not found or could not be run. You might need to set the GST_PLUGIN_SCANNER environment variable if your setup is unusual. This should normally not be required though.
  │ 0:00:00.016606787 32203 0x55b9f69d52a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libasound.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:32203): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstalsa.so': libasound.so.2: cannot open shared object file: No such file or directory
  │ 0:00:00.016898248 32203 0x55b9f69d52a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libmms.so.0: cannot open shared object file: No such file or directory
  │ 
  │ (servo:32203): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstmms.so': libmms.so.0: cannot open shared object file: No such file or directory
  │ 0:00:00.017039791 32203 0x55b9f69d52a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libfaad.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:32203): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstfaad.so': libfaad.so.2: cannot open shared object file: No such file or directory
  │ called `Option::unwrap()` on a `None` value (thread WebGLThread, at src/libcore/option.rs:347)
  │ stack backtrace:
  │    0: servo::main::{{closure}}::h2778f06a7b19c5ed (0x55b9f0d7b16f)
  │    1: std::panicking::rust_panic_with_hook::h94a43555bff30e2e (0x55b9f404f508)
  │              at src/libstd/panicking.rs:479
  │    2: std::panicking::continue_panic_fmt::h6db1fa8c1d9cf848 (0x55b9f404efa1)
  │              at src/libstd/panicking.rs:382
  │    3: rust_begin_unwind (0x55b9f404ee85)
  │              at src/libstd/panicking.rs:309
  │    4: core::panicking::panic_fmt::h00d836662640edaa (0x55b9f40721dc)
  │              at src/libcore/panicking.rs:85
  │    5: core::panicking::panic::h33a036b88924a6d5 (0x55b9f407211b)
  │              at src/libcore/panicking.rs:49
  │    6: canvas::webgl_thread::WebGLThread<VR>::handle_webgl_command::hdb5ed964ce6dbc57 (0x55b9f337e097)
  │    7: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h392b0167bb7765b8E.llvm.324392017536448607 (0x55b9f33637c3)
  │    8: _ZN3std9panicking3try7do_call17hd00c52b924187308E.llvm.12642146261625164759 (0x55b9f33c2958)
  │    9: __rust_maybe_catch_panic (0x55b9f4059629)
  │              at src/libpanic_unwind/lib.rs:82
  │   10: core::ops::function::FnOnce::call_once{{vtable.shim}}::h8c071aff7a1d75e2 (0x55b9f3367644)
  │   11: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::he874ac8516b9b7f7 (0x55b9f403e04e)
  │              at /rustc/04a3dd8a872633ca1e4c217d11f741cc35cb19a5/src/liballoc/boxed.rs:746
  │   12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb961fc621fe92936 (0x55b9f40589af)
  │              at /rustc/04a3dd8a872633ca1e4c217d11f741cc35cb19a5/src/liballoc/boxed.rs:746
  │       std::sys_common::thread::start_thread::h616ece2fd1361d9e
  │              at src/libstd/sys_common/thread.rs:13
  │       std::sys::unix::thread::Thread::new::thread_start::h1cad89e856422f4e
  │              at src/libstd/sys/unix/thread.rs:79
  │   13: start_thread (0x7f4c0db3a6b9)
  │   14: clone (0x7f4c0c1b341c)
  │   15: <unknown> (0x0)

  ▶ CRASH [expected OK] /_webgl/conformance/extensions/oes-standard-derivatives.html
  │ 
  │ VMware, Inc.
  │ softpipe
  │ 3.3 (Core Profile) Mesa 18.3.0-devel
  │ 
  │ (servo:92702): GStreamer-WARNING **: External plugin loader failed. This most likely means that the plugin loader helper binary was not found or could not be run. You might need to set the GST_PLUGIN_SCANNER environment variable if your setup is unusual. This should normally not be required though.
  │ 0:00:00.021898826 92702 0x55edd59562a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libasound.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:92702): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstalsa.so': libasound.so.2: cannot open shared object file: No such file or directory
  │ 0:00:00.022286537 92702 0x55edd59562a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libmms.so.0: cannot open shared object file: No such file or directory
  │ 
  │ (servo:92702): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstmms.so': libmms.so.0: cannot open shared object file: No such file or directory
  │ 0:00:00.022518071 92702 0x55edd59562a0 WARN      GST_PLUGIN_LOADING gstplugin.c:793:_priv_gst_plugin_load_file_for_registry: module_open failed: libfaad.so.2: cannot open shared object file: No such file or directory
  │ 
  │ (servo:92702): GStreamer-WARNING **: Failed to load plugin '/home/servo/gst/lib/gstreamer-1.0/libgstfaad.so': libfaad.so.2: cannot open shared object file: No such file or directory
  │ called `Option::unwrap()` on a `None` value (thread WebGLThread, at src/libcore/option.rs:347)
  │ stack backtrace:
  │    0: servo::main::{{closure}}::h2778f06a7b19c5ed (0x55edd02b816f)
  │    1: std::panicking::rust_panic_with_hook::h94a43555bff30e2e (0x55edd358c508)
  │              at src/libstd/panicking.rs:479
  │    2: std::panicking::continue_panic_fmt::h6db1fa8c1d9cf848 (0x55edd358bfa1)
  │              at src/libstd/panicking.rs:382
  │    3: rust_begin_unwind (0x55edd358be85)
  │              at src/libstd/panicking.rs:309
  │    4: core::panicking::panic_fmt::h00d836662640edaa (0x55edd35af1dc)
  │              at src/libcore/panicking.rs:85
  │    5: core::panicking::panic::h33a036b88924a6d5 (0x55edd35af11b)
  │              at src/libcore/panicking.rs:49
  │    6: canvas::webgl_thread::WebGLThread<VR>::handle_webgl_command::hdb5ed964ce6dbc57 (0x55edd28bb097)
  │    7: _ZN3std10sys_common9backtrace28__rust_begin_short_backtrace17h392b0167bb7765b8E.llvm.324392017536448607 (0x55edd28a07c3)
  │    8: _ZN3std9panicking3try7do_call17hd00c52b924187308E.llvm.12642146261625164759 (0x55edd28ff958)
  │    9: __rust_maybe_catch_panic (0x55edd3596629)
  │              at src/libpanic_unwind/lib.rs:82
  │   10: core::ops::function::FnOnce::call_once{{vtable.shim}}::h8c071aff7a1d75e2 (0x55edd28a4644)
  │   11: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::he874ac8516b9b7f7 (0x55edd357b04e)
  │              at /rustc/04a3dd8a872633ca1e4c217d11f741cc35cb19a5/src/liballoc/boxed.rs:746
  │   12: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once::hb961fc621fe92936 (0x55edd35959af)
  │              at /rustc/04a3dd8a872633ca1e4c217d11f741cc35cb19a5/src/liballoc/boxed.rs:746
  │       std::sys_common::thread::start_thread::h616ece2fd1361d9e
  │              at src/libstd/sys_common/thread.rs:13
  │       std::sys::unix::thread::Thread::new::thread_start::h1cad89e856422f4e
  │              at src/libstd/sys/unix/thread.rs:79
  │   13: start_thread (0x7f4f871286b9)
  │   14: clone (0x7f4f857a141c)
  │   15: <unknown> (0x0)
```

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23696) made this pull request unmergeable. Please resolve the merge conflicts.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23827) made this pull request unmergeable. Please resolve the merge conflicts.

jdm commented on components/canvas/gl_context.rs:
```diff
@@ -168,44 +193,101 @@ impl GLContextWrapper {
                     let draw_buffer = ctx.borrow_draw_buffer().unwrap();
                     (
                         draw_buffer.size(),
                         draw_buffer.get_bound_texture_id().unwrap(),
```
If we make get_bound_texture_id return a TextureBacking object, we don't need to create a separate `io_surface_id` value.

jdm commented on components/canvas/gl_context.rs:
```diff
@@ -168,44 +193,101 @@ impl GLContextWrapper {
     }
 
-    pub fn get_info(&self) -> (Size2D<i32>, u32, GLLimits) {
+    pub fn get_info(&self) -> (Size2D<i32>, u32, Option<u32>, GLLimits) {
```
I think we can have get_info return `(Size2D<i32>, TextureBacking, GLLimits)` instead.

jdm commented on components/canvas/gl_context.rs:
```diff
@@ -168,44 +193,101 @@ impl GLContextWrapper {
+        &mut self,
+        _clear_color: (f32, f32, f32, f32),
+        _mask: u32,
+    ) -> Option<u32> {
```
I think we'll want to return Option<TextureBacking> here.

jdm commented on components/canvas/webgl_thread.rs:
```diff
@@ -763,6 +825,21 @@ struct WebGLContextInfo {
     /// The status of this context with respect to external consumers.
     render_state: ContextRenderState,
+    /// The ID of the IOSurface which we can send to the WR thread
+    io_surface_id: Option<u32>,
```
I think we don't need this if we make texture_id a TextureBacking value.

jdm commented on components/canvas/gl_context.rs:
```diff
@@ -168,44 +193,101 @@ impl GLContextWrapper {
+        }
+    }
+
+    pub fn get_active_io_surface_id(&self) -> Option<u32> {
```
I think we can get rid of this with my proposed changes.

jdm commented on components/canvas/webgl_thread.rs:
```diff
@@ -273,8 +314,13 @@ impl<VR: WebVRRenderHandler + 'static> WebGLThread<VR> {
+        if !info.has_request_animation {
+            info.io_surface_id = data.ctx.get_active_io_surface_id();
+        } else {
+            info.io_surface_id = data.ctx.handle_lock();
```
This is one place where it is not clear to me if my plan to only store a single TextureBacking value in WebGLContextInfo will be enough.

jdm commented on components/script/dom/window.rs:
```diff
@@ -893,7 +894,14 @@ impl WindowMethods for Window {
-            .request_animation_frame(AnimationFrameCallback::FrameRequestCallback { callback })
+            .request_animation_frame(AnimationFrameCallback::FrameRequestCallback {
+                callback,
+                webgl_chan: if cfg!(target_os = "macos") && opts::get().with_io_surface {
```
We should do this unconditionally so it's easier to reason about the implementation. We can make the swap operation in the backend a no-op if necessary.

jdm commented on components/canvas/webgl_mode/inprocess.rs:
```diff
@@ -73,18 +77,29 @@ impl WebGLThreads {
+impl Drop for WebGLExternalImages {
+    fn drop(&mut self) {
+        for (_, texture_id) in &self.textures {
+            self.webrender_gl.delete_textures(&[*texture_id]);
```
We will want to have a followup issue filed for deleting these textures when the corresponding webgl context is removed.

jdm commented on components/canvas/gl_context.rs:
```diff
@@ -141,6 +160,8 @@ impl GLContextFactory {
     Native(GLContext<NativeGLContext>),
     OSMesa(GLContext<OSMesaContext>),
+    #[cfg(target_os = "macos")]
+    NativeWithIOSurface(GLContext<NativeGLContext>),
```
My overall goal in my suggestions is to remove this new enum variant and move the platform-specific behaviour either into rust-offscreen-rendering-context by adding abstractions like TextureBacking, or use match statements on TextureBacking values in this code to choose which behaviour to use.

jdm reviewed (changes requested):
Ok, here are some initial thoughts on how we can integrate this work. Ideally, I would like to have a single codepath where the code in canvas/ is written using the triple buffering scheme everywhere, and there are a few platform-specific bits where we need to do special work with IOSurface values in particular. If there's a cross-platform API for triple-buffering in rust-offscreen-rendering-context that uses TextureBacking values everywhere, it can still have no-op implementations for the non-iosurface code paths, and it will be easier to reason about the intention of the code.

pcwalton commented:
`io_surface::lookup()` is a security problem. We want to kill usage of this.

zakorgy commented:
> `io_surface::lookup()` is a security problem. We want to kill usage of this.

Thanks @pcwalton for warning us. Can we avoid this by using one of the IOSurfaceCreateMachPort or IOSurfaceCreateXPCObject?

Commit:
Add TextureBacking abstarction above textures and IOSurfaces

Commit:
Remove NativeWithIOSurface GLContextWrapper variant

zakorgy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/servo/servo/23509)* status: 0 of 11 files reviewed, 17 unresolved discussions (waiting on @asajeffrey, @jdm, and @zakorgy)

___
*[components/canvas/gl_context.rs, line 164 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6S-r11-164:-LlaaE0d7wFSqESOEMP2:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/gl_context.rs#L164)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

My overall goal in my suggestions is to remove this new enum variant and move the platform-specific behaviour either into rust-offscreen-rendering-context by adding abstractions like TextureBacking, or use match statements on TextureBacking values in this code to choose which behaviour to use.
</blockquote></details>

Done.
___
*[components/canvas/gl_context.rs, line 212 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6S-r11-212:-LlaaGB34RIRE5YnozII:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/gl_context.rs#L212)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

I think we can have get_info return `(Size2D<i32>, TextureBacking, GLLimits)` instead.
</blockquote></details>

Done.
___
*[components/canvas/gl_context.rs, line 219 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6S-r11-219:-LlaaMTy6XRvTYIPh74U:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/gl_context.rs#L219)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

If we make get_bound_texture_id return a TextureBacking object, we don't need to create a separate `io_surface_id` value.
</blockquote></details>

Done.
___
*[components/canvas/gl_context.rs, line 260 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6S-r11-260:-LlaaO2005WkYmMtItqu:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/gl_context.rs#L260)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

I think we can get rid of this with my proposed changes.
</blockquote></details>

Done.
___
*[components/canvas/gl_context.rs, line 276 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6S-r11-276:-LlaaTR95skKueI1yCjt:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/gl_context.rs#L276)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

I think we'll want to return Option<TextureBacking> here.
</blockquote></details>

Done.
___
*[components/canvas/webgl_thread.rs, line 321 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6U-r11-321:-LlabLU56pJALir_OaZl:b-qlkcj5) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/webgl_thread.rs#L321)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

This is one place where it is not clear to me if my plan to only store a single TextureBacking value in WebGLContextInfo will be enough.
</blockquote></details>

It can work with a single `TextureBacking` value.
___
*[components/canvas/webgl_thread.rs, line 829 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6U-r11-829:-Llabf5E-TrGGnzlS0wy:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/canvas/webgl_thread.rs#L829)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

I think we don't need this if we make texture_id a TextureBacking value.
</blockquote></details>

Done.
___
*[components/script/dom/window.rs, line 899 at r11](https://reviewable.io/reviews/servo/servo/23509#-LgXtUGiNg3w7fI3pO6Y-r11-899:-LlacqkIDKCk73G_JtIf:b-896fix) ([raw file](https://github.com/servo/servo/blob/cc86944f3a5e854c4852692781da59e734d8790d/components/script/dom/window.rs#L899)):*
<details><summary><i>Previously, jdm (Josh Matthews) wrote…</i></summary><blockquote>

We should do this unconditionally so it's easier to reason about the implementation. We can make the swap operation in the backend a no-op if necessary.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

zakorgy commented:
@jdm I have updated the code as your described, and removed the `NativeWithIOSurface` variant. I've used your `offscren_gl_context` branch with a minimal [modification](https://github.com/zakorgy/rust-offscreen-rendering-context/commit/02345d40c5c3e5b3024664775494f8c69b08ec18).

> `io_surface::lookup()` is a security problem. We want to kill usage of this.

I will check how we could use `IOSurfaceCreateMachPort` or `IOSurfaceCreateXPCObject` and their lookup pairs instead of the current one.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #23927) made this pull request unmergeable. Please resolve the merge conflicts.
